Update the "open a file" dialog to use promises (doesn't take callbacks as of Electron 7) (for Electron 9 PR)

DeeDeeG opened the pull request:
<details><summary>Requirements for Contributing a Bug Fix (from template, click to expand):</summary>

* Fill out the template below. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* The pull request must only fix an existing bug. To contribute other changes, you must use a different template. You can see all templates at <https://github.com/atom/.github/tree/master/.github/PULL_REQUEST_TEMPLATE>.
* The pull request must update the test suite to demonstrate the changed functionality. For guidance, please see <https://flight-manual.atom.io/hacking-atom/sections/writing-specs/>.
* After you create the pull request, all status checks must be pass before a maintainer reviews your contribution. For more details, please see <https://github.com/atom/.github/tree/master/CONTRIBUTING.md#pull-requests>.

</details>

### Identify the Bug

<!--

Link to the issue describing the bug that you're fixing.

If there is not yet an issue for your bug, please open a new issue and then link to that issue in your pull request.
Note: In some cases, one person's "bug" is another person's "feature." If the pull request does not address an existing issue with the "bug" label, the maintainers have the final say on whether the current behavior is a bug.

-->

Can't open projects in the Electron 9 build of Atom.

- https://github.com/atom/atom/pull/21815#issuecomment-759780153
- https://github.com/atom/atom/pull/21777#issuecomment-747027422

### Description of the Change

<!--

We must be able to understand the design of your change from this description. If we can't get a good idea of what the code will be doing from the description here, the pull request may be closed at the maintainers' discretion. Keep in mind that the maintainer reviewing this PR may not be familiar with or have worked with the code here recently, so please walk us through the concepts.

-->

Use Promises instead of callbacks with the `dialog.showOpenDialog()` Electron API call.

Why? Because the Electron API changed. Many API calls don't accept callbacks anymore, and are meant to be used with Promises now. This happened starting in Electron 7.0.0, apparently.

See: 
- https://github.com/electron/electron/pull/17907
  - A big PR that removed lots of (previously deprecated) callback-based async APIs in favor of Promises, with separate and more-explicitly Sync versions available for some of the API calls. This landed in Electron 7.
- https://www.electronjs.org/docs/api/modernization/promisification
  - Overall documentation of moving many API calls to Promises

### Alternate Designs

<!-- Explain what other alternates were considered and why the proposed version was selected -->

None. But please review the specifics of this. I'm new to using Promises.

### Possible Drawbacks

<!-- What are the possible side-effects or negative impacts of the code change? -->

None. (But please review that I did this right.)

### Verification Process

<!--

What process did you follow to verify that the change has not introduced any regressions? Describe the actions you performed (including buttons you clicked, text you typed, commands you ran, etc.), and describe the results you observed.

-->

I ran this change in dev mode using a build of the Electron 9 PR branch. I manually verified that projects can be added and files can be opened now using the "open a file" dialog.

Tested on macOS 10.15.7 Catalina.

### Release Notes

<!--

Please describe the changes in a single line that explains this improvement in
terms that a user can understand. This text will be used in Atom's release notes.

If this change is not user-facing or notable enough to be included in release notes
you may use the strings "Not applicable" or "N/A" here.

Examples:

- The GitHub package now allows you to add co-authors to commits.
- Fixed an issue where multiple cursors did not work in a file with a single line.
- Increased the performance of searching and replacing across a whole project.

-->

N/A

DeeDeeG commented:
I'm not sure if this changes any public Atom API behavior at all, but if it does, I'd like to put some release notes for this PR instead of putting `N/A`. If anyone knows whether this changes public APIs, I'd appreciate that info so I can add an appropriate release notes entry.

DeeDeeG commented:
I see this PR didn't run any CI because it doesn't target the `master` branch. I'm doing a CI run at my fork:
https://dev.azure.com/DeeDeeG/b/_build/results?buildId=1082

DeeDeeG commented:
I broke some tests with this, apparently. Looking into it...

asturur commented:
i was working exactly on this showOpenDialog, should i take another part? do we want to sync on a list?

DeeDeeG commented:
@asturur I got a bit stuck on this, since a few of the tests are broken. I found the tests much more confusing to fix than the app itself. (Although the broken tests make me less certain I fixed the app 100% correctly... It worked in manual testing.)

@asturur I'm not sure how different our code looks for these things, so I'd be curious to see what your fix looked like and compare.

But my curiosity aside, if you have a version with working tests and a correct fix, that would be better than what I have here. Thanks for working on this.

> sync on a list

If I start working on another one I'll `@mention` you and try to sync up. (I think I want to make sure the "open file" dialog is fixed before moving on to any other API calls, personally.)

asturur commented on src/main-process/atom-application.js:
```diff
@@ -1946,76 +1946,75 @@ module.exports = class AtomApplication extends EventEmitter {
-    const openOptions = {
-      properties: properties.concat(['multiSelections', 'createDirectory']),
-      title: (() => {
+  promptForPath(type, path) {
```
i m not sure how you develop a plugin for atom, but isn't there the risk that something outside atom could use this promptForPath and the signature just changed?

DeeDeeG replied:
If I were to try this again, I would try to copy your approach from https://github.com/atom/atom/pull/21934 / https://github.com/atom/atom/pull/21935.

That seems to be a simpler change with less side-effects.

Your approach would also probably not require me to change any of the tests! I'd like to wait and see what happens with those two Pull Requests. And I'll try to copy the final result from there.

---

If you want to go ahead and do this yourself I wouldn't stop you. I definitely wanted to report the problem, which I did in https://github.com/atom/atom/pull/21777#issuecomment-773014459. And I wanted to attempt this myself, both to learn how to use Promises, and to "get the ball rolling" and start the process, so other people would be more likely to act on it. As it turns out, I have a bit more to learn about Promises, function signatures... and so on.

sadick254 replied:
I have merged the 2 PRs #21934 and  #21935. It would be good if the approach on those 2 PR's are followed. @DeeDeeG Let me know when this PR is ready for review.

DeeDeeG replied:
@sadick254 your commit https://github.com/atom/atom/pull/21777/commits/e5267bcade57ffdb40e39e5cf5910eb71d43365d from the Electron 9 PR looks right.

I expect that commit works perfectly. But I'll do a quick CI run with that against `master` to confirm and see how it goes.

DeeDeeG replied:
Okay, I am working on this.

Two things:

1) The Promise gives an object back:

```js
{
  canceled: true|false // Boolean - whether or not the dialog was canceled.
  filePaths: [''] // String[] - An array of file paths chosen by the user. If the dialog is cancelled this will be an empty array.
  bookmarks: [''] // something specific to Mac App Store builds
}
```

So I believe the callback should be run with those as arguments like in @asturur's PR, if the callback gets called at all.

2) Might need to conditionally do `.then(callback)` only if `typeof callback === 'function'` like in @asturur's PR.

DeeDeeG commented:
Note to maintainers: This was a first attempt, but it is not in good shape to be merged. Please do not merge this as it is right now, thank you.

asturur commented:
Aren't other api to look at? i got to half list and i could find only open/save dialog, but i still miss other half of the promisification list from electron. I ended up short on time at the end.

DeeDeeG committed:
Update dialog.showOpenDialog() to use Promises

Fixes the "Open a file" and "Add a project" workflows
for Electron 7 and newer, where the function
no-longer accepts/runs callbacks passed to it.

See this documentation:

https://github.com/electron/electron/blob/v9.4.3/docs/api/modernization/promisification.md
https://github.com/electron/electron/blob/master/docs/api/dialog.md#dialogshowopendialogbrowserwindow-options

This commit is basically copy-pasting the same solution
implemented by @asturur in this pull request:
https://github.com/atom/atom/pull/21935

Co-authored-by: Andrea Bogazzi <andreabogazzi79@gmail.com>

DeeDeeG commented:
@asturur there are severa updated API functions in the promisification document:

https://github.com/electron/electron/blob/v9.4.3/docs/api/modernization/promisification.md

But Atom does not use all of those. I mentioned `dialog.showOpenDialog()` and `dialog.showSaveDialog()` as soon as I knew they needed to be updated. So next would be to keep going past those on the list and do a code search for all of those.

I intend to do that once these two are both fixed/merged.

DeeDeeG commented:
@sadick254 I updated this PR to copy the approach from https://github.com/atom/atom/pull/21935.

I think this is more correct.

The tests pass with https://github.com/atom/atom/commit/e5267bcade57ffdb40e39e5cf5910eb71d43365d, but the actual "Open a file" dialog doesn't work for me in dev mode with that commit. The commit I just pushed (https://github.com/atom/atom/commit/becb57b626a867389e56674f09c604363cb91348) works for me manually testing in dev mode.

And I am running CI for this branch: https://dev.azure.com/DeeDeeG/b/_build/results?buildId=1089&view=results

asturur commented:
> @asturur there are several updated API functions in the promisification document:
> 
> https://github.com/electron/electron/blob/v9.4.3/docs/api/modernization/promisification.md
> 
> But Atom does not use all of those. I mentioned `dialog.showOpenDialog()` and `dialog.showSaveDialog()` as soon as I knew they needed to be updated. So next would be to keep going past those on the list and do a code search for all of those.
> 
> (For example: `cd atom` and `git grep function_name_here` in the Atom repo.)
> 
> I intend to do that once these two are both fixed/merged.

so i ll resume from end of the list and i ll go up. First half of the list is unused by atom.

DeeDeeG commented:
Interesting historical note (no action needed): We do use `dialog.showMessageBox()` and `dialog.showMessageBoxSync()`, but this transition was already handled in some other PRs: https://github.com/atom/atom/pull/16229 https://github.com/atom/atom/pull/21587

DeeDeeG commented:
We appear to use these (based on a quick code search), and I've listed what I think we need to do about them, if anything:
- [`ses.resolveProxy(url, callback)`](https://github.com/electron/electron/blob/v5.0.13/docs/api/session.md#sesresolveproxyurl-callback) _becomes - - >_ [`ses.resolveProxy(url)`](https://github.com/electron/electron/blob/v9.4.3/docs/api/session.md#sesresolveproxyurl)
  - I think we need to fix this one. See [`ses.resolveProxy(url, callback)`](https://github.com/electron/electron/blob/v6.1.12/docs/api/session.md#sesresolveproxyurl-callback)
    - > Resolves the proxy information for `url`. The `callback` will be called with `callback(proxy)` when the request is performed.
- [`shell.openExternal(url[, options, callback])`](https://github.com/electron/electron/blob/v4.2.12/docs/api/shell.md#shellopenexternalurl-options-callback) _becomes - - >_ [`shell.openExternal(url[, options])`](https://github.com/electron/electron/blob/v9.4.3/docs/api/shell.md#shellopenexternalurl-options)
  - I think this is fine, because we're not passing callbacks.
- [`contents.executeJavaScript(code[, userGesture, callback])`](https://github.com/electron/electron/blob/v5.0.13/docs/api/web-contents.md#contentsexecutejavascriptcode-usergesture-callback) _becomes - - >_ [`contents.executeJavaScript(code[, userGesture])`](https://github.com/electron/electron/blob/v9.4.3/docs/api/web-contents.md#contentsexecutejavascriptcode-usergesture)
  - I think this is fine, because we're not passing callbacks.

ThatXliner commented:
### I'm no JavaScript expert

But can't you just shove a `.then(callback)`?

DeeDeeG commented:
@ThatXliner I'm also not a JavaScript expert, and yes that's the fix I'm trying. 😆

DeeDeeG commented:
I'm closing this PR. @sadick254 has added this to the Electron 9 PR as commit https://github.com/atom/atom/pull/21777/commits/0a3e76284d4236aaf3f18b4344b29691867b77bd.
